modules/standard/Spawn.chpl

```diff
@@ -0,0 +1,336 @@
+      ret.stdout_pipe = true;
+      ret.stdout_file = openfd(stdout_fd, error=err, hints=QIO_HINT_OWNED);
+      if err {
+        ret.spawn_error = err; return ret;
+      }
+
+      ret.stdout = ret.stdout_file.reader(error=err);
+      if err {
+        ret.spawn_error = err; return ret;
+      }
+    }
+
+    if stderr_pipe {
+      ret.stderr_pipe = true;
+      ret.stderr_file = openfd(stderr_fd, error=err, hints=QIO_HINT_OWNED);
+      if err {
+        ret.spawn_error = err; return ret;
+      }
+
+      ret.stderr = ret.stderr_file.reader(error=err);
+      if err {
+        ret.spawn_error = err; return ret;
+      }
+    }
+
+    return ret;
+  }
+
+  proc spawnshell(command:string, env:[] string=Spawn.empty_env,
+                  stdin:?t = FORWARD, stdout:?u = FORWARD, stderr:?v = FORWARD,
+                  executable="/bin/sh", shellarg="-c",
+                  param kind=iokind.dynamic, param locking=true)
+  {
+    var args = [command];
+    if shellarg != "" then args.push_front(shellarg);
+    args.push_front(executable);
+    return spawn(args, env, executable, kind=kind, locking=locking);
+  }
+
+  proc subprocess.poll(out error:syserr) {
+    on home {
+      // check if child process has terminated.
+      var done:c_int = 0;
+      var exitcode:c_int = 0;
+      error = qio_waitpid(pid, 0, done, exitcode);
+      if done {
+        this.running = false;
+        this.exit_status = exitcode;
+      }
+    }
+  }
+  proc subprocess.wait(out error:syserr) {
+
+    error = this.spawn_error;
+    if !running then return;
+
+    on home {
+      
+      // Close stdin.
+      if this.stdin_pipe {
+        // send data to stdin
+        if this.stdin_buffering then this.stdin._commit();
+        this.stdin.close(error=error);
+        this.stdin_file.close(error=error);
+      }
+
+      // check if child process has terminated.
+      var done:c_int = 0;
+      var exitcode:c_int = 0;
+      error = qio_waitpid(pid, 1, done, exitcode);
+      if done {
+        this.running = false;
+        this.exit_status = exitcode;
+      }
+    }
+  }
+  proc subprocess.wait() {
+    var err:syserr = ENOERR;
+
+    this.wait(error=err);
+    if err then halt("Error in subprocess.wait: " + errorToString(err));
+  }
+
+  // wait for process to terminate...
+  // commit stdin, sending any data to the process,
+  // while also reading any output from the process
+  // into the stdout and stderr channel buffers.
+  proc subprocess.communicate(out error:syserr) {
+    
+    on home {
+      stdin._commit();
+
+      error = qio_proc_communicate(
+        locking, 
+        stdin._channel_internal,
+        stdout._channel_internal,
+        stderr._channel_internal);
+      /*
+        if stdin_pipe then stdin._channel_internal else QIO_CHANNEL_PTR_NULL,
+        if stdout_pipe then stdout._channel_internal else QIO_CHANNEL_PTR_NULL,
+        if stderr_pipe then stderr._channel_internal else
+        QIO_CHANNEL_PTR_NULL);*/
```

Comment:
Is this intended to explain the potential states of the _channel_internal fields, or is it old code that might need to be cleaned up?